node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainer.java

```diff
@@ -79,33 +82,68 @@ protected double maintain() {
             return 0;  // avoid removing excess hosts
         }
 
-        return markForRemoval(excessHosts);
+        return markForRemoval(provisionedSnapshot);
     }
 
-    private double markForRemoval(List<Node> excessHosts) {
-        if (excessHosts.isEmpty()) return 1;
+    private double markForRemoval(List<Node> provisionedSnapshot) {
+        // Group nodes by parent; no parent means it's a host.
+        Map<Optional<String>, List<Node>> nodesByParent = provisionedSnapshot.stream().collect(groupingBy(Node::parentHostname));
+
+        // Find all hosts that we once thought were empty (first clouse), or whose children are now all removable (second clause).
+        List<Node> emptyHosts = nodesByParent.get(Optional.<String>empty()).stream()
+                                             .filter(host ->    host.hostEmptyAt().isPresent()
+                                                             || nodesByParent.getOrDefault(Optional.of(host.hostname()), List.of())
+                                                                             .stream().allMatch(HostCapacityMaintainer::canDeprovision))
+                                             .toList();
+
+        if (emptyHosts.isEmpty()) return 1;
 
         int attempts = 0, success = 0;
-        for (List<Node> typeExcessHosts : excessHosts.stream().collect(Collectors.groupingBy(Node::type)).values()) {
+        for (Set<Node> typeEmptyHosts : emptyHosts.stream().collect(groupingBy(Node::type, toSet())).values()) {
             attempts++;
             // All nodes in the list are hosts of the same type, so they use the same lock regardless of their allocation
-            Optional<NodeMutex> appMutex = nodeRepository().nodes().lockAndGet(typeExcessHosts.get(0), Duration.ofSeconds(10));
+            Optional<NodeMutex> appMutex = nodeRepository().nodes().lockAndGet(typeEmptyHosts.iterator().next(), Duration.ofSeconds(10));
             if (appMutex.isEmpty()) continue;
             try (Mutex lock = appMutex.get();
                  Mutex unallocatedLock = nodeRepository().nodes().lockUnallocated()) {
                 // Re-read all nodes under lock and compute the candidates for removal. The actual nodes we want
-                // to mark for removal is the intersection with typeExcessHosts
-                List<Node> toMarkForRemoval = candidatesForRemoval(nodeRepository().nodes().list().asList()).stream()
-                        .filter(typeExcessHosts::contains)
-                        .toList();
+                // to mark for removal is the intersection with typeEmptyHosts, which excludes the preprovisioned hosts.
+                Map<Optional<String>, List<Node>> currentNodesByParent = nodeRepository().nodes().list().stream().collect(groupingBy(Node::parentHostname));
+                List<Node> candidateHosts = new ArrayList<>(currentNodesByParent.get(Optional.<String>empty()));
+                candidateHosts.retainAll(typeEmptyHosts);
```

Comment:
Hmm, why, though? We're only getting false negatives, which isn't a problem, since this is a maintainer?
To do that, we'd have to somehow combine the newly provisioned part of the provisionedSnapshot, with the nodes we read from the node repository. Is it easy to know whether a node was provisioned by us, now, or just vanished from the node repository in the meantime?